// @angular/compiler/src/private_export.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
"use strict";
var directive_normalizer = require('./directive_normalizer');
var lexer = require('./expression_parser/lexer');
var parser = require('./expression_parser/parser');
var metadata_resolver = require('./metadata_resolver');
var html_parser = require('./ml_parser/html_parser');
var interpolation_config = require('./ml_parser/interpolation_config');
var ng_module_compiler = require('./ng_module_compiler');
var path_util = require('./output/path_util');
var ts_emitter = require('./output/ts_emitter');
var parse_util = require('./parse_util');
var dom_element_schema_registry = require('./schema/dom_element_schema_registry');
var selector = require('./selector');
var style_compiler = require('./style_compiler');
var template_parser = require('./template_parser/template_parser');
var view_compiler = require('./view_compiler/view_compiler');
exports.__compiler_private__ = {
    SelectorMatcher: selector.SelectorMatcher,
    CssSelector: selector.CssSelector,
    AssetUrl: path_util.AssetUrl,
    ImportGenerator: path_util.ImportGenerator,
    CompileMetadataResolver: metadata_resolver.CompileMetadataResolver,
    HtmlParser: html_parser.HtmlParser,
    InterpolationConfig: interpolation_config.InterpolationConfig,
    DirectiveNormalizer: directive_normalizer.DirectiveNormalizer,
    Lexer: lexer.Lexer,
    Parser: parser.Parser,
    ParseLocation: parse_util.ParseLocation,
    ParseError: parse_util.ParseError,
    ParseErrorLevel: parse_util.ParseErrorLevel,
    ParseSourceFile: parse_util.ParseSourceFile,
    ParseSourceSpan: parse_util.ParseSourceSpan,
    TemplateParser: template_parser.TemplateParser,
    DomElementSchemaRegistry: dom_element_schema_registry.DomElementSchemaRegistry,
    StyleCompiler: style_compiler.StyleCompiler,
    ViewCompiler: view_compiler.ViewCompiler,
    NgModuleCompiler: ng_module_compiler.NgModuleCompiler,
    TypeScriptEmitter: ts_emitter.TypeScriptEmitter
};
